src/Native/Unix/System.Native/pal_io.c

```diff
@@ -1188,30 +1192,88 @@ static int32_t CopyFile_ReadWrite(int inFd, int outFd)
 }
 #endif // !HAVE_FCOPYFILE
 
-int32_t SystemNative_CopyFile(intptr_t sourceFd, intptr_t destinationFd)
+int32_t SystemNative_CopyFile(intptr_t sourceFd, const char* srcPath, const char* destPath, int32_t overwrite)
 {
     int inFd = ToFileDescriptor(sourceFd);
-    int outFd = ToFileDescriptor(destinationFd);
-
-#if HAVE_FCOPYFILE
-    // If fcopyfile is available (OS X), try to use it, as the whole copy
-    // can be performed in the kernel, without lots of unnecessary copying.
-    // Copy data and metadata.
-    return fcopyfile(inFd, outFd, NULL, COPYFILE_ALL) == 0 ? 0 : -1;
-#else
-    // Get the stats on the source file.
+    int outFd;
     int ret;
     struct stat_ sourceStat;
-    bool copied = false;
-#if HAVE_SENDFILE_4
-    // If sendfile is available (Linux), try to use it, as the whole copy
-    // can be performed in the kernel, without lots of unnecessary copying.
+
     while ((ret = fstat_(inFd, &sourceStat)) < 0 && errno == EINTR);
     if (ret != 0)
     {
         return -1;
     }
 
+    struct stat_ destStat;
+    while ((ret = stat_(destPath, &destStat)) < 0 && errno == EINTR);
+    if (ret == 0)
+    {
+        if (sourceStat.st_dev == destStat.st_dev && sourceStat.st_ino == destStat.st_ino)
+        {
+            errno = EBUSY;
+            return -1;
+        }
+
+        if (!overwrite)
+        {
+            errno = EEXIST;
+            return -1;
+        }
+
+#if HAVE_CLONEFILE
+        // For clonefile we need to unlink the destination file first but we need to
+        // check permission first to ensure we don't try to unlink read-only file
+        if (access(destPath, W_OK) != 0)
+        {
+            errno = EACCES;
+            return -1;
+        }
+        
+        ret = unlink(destPath);
+        if (ret != 0)
+        {
+            return ret;
+        }
+#endif
+    }
+
+#if HAVE_CLONEFILE
+    while ((ret = clonefile(srcPath, destPath, 0)) < 0 && errno == EINTR);
+    // EEXIST can happen due to race condition between the stat/unlink above
+    // and the clonefile here. The file could be (re-)created from another
+    // thread or process before we have a chance to call clonefile. Handle
+    // it by falling back to the slow path.
+    if (ret == 0 || (errno != ENOTSUP && errno != EXDEV && errno != EEXIST))
+    {
+        return ret;
+    }
+#else
+    // Unused variable
+    (void)srcPath;
+#endif
+
+    while ((outFd = open(destPath, O_WRONLY | O_TRUNC | O_CREAT | (overwrite ? 0 : O_EXCL), sourceStat.st_mode & (S_IRWXU | S_IRWXG | S_IRWXO))) < 0 && errno == EINTR);
```

Comment:
Can/should we make the fd close-on-exec?

Reply:
Can you think of any scenario where would that matter? I used the same flags as the previous code did (on the managed side) so this is not changing the behavior.

Reply:
Process.Start concurrently with File.Copy, resulting in the child process keeping the destination file open?

Reply:
> I used the same flags as the previous code did (on the managed side) so this is not changing the behavior

FileStream was being used before, and it defaults to non-inheritable:
https://github.com/dotnet/coreclr/blob/25520b174be2a8584e9d016a1eaa06b6120c8d6a/src/System.Private.CoreLib/shared/System/IO/FileStream.Unix.cs#L180-L184

Reply:
Ah, I stand corrected. I'll add it then.